src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs

```diff
@@ -0,0 +1,1317 @@
+        // all come from entrypoints that are specific to a particular project.
+        return Branch(newSolutionState);
+    }
+
+    /// <inheritdoc cref="SolutionState.RemoveAnalyzerReference(ProjectId, AnalyzerReference)"/>
+    public SolutionCompilationState RemoveAnalyzerReference(StateChange stateChange, AnalyzerReference analyzerReference)
+    {
+        return ForkProject(
+            stateChange,
+            static (stateChange, analyzerReference) => new CompilationAndGeneratorDriverTranslationAction.AddOrRemoveAnalyzerReferencesAction(
+                stateChange.OldProjectState.Language, referencesToRemove: ImmutableArray.Create(analyzerReference)),
+            forkTracker: true,
+            arg: analyzerReference);
+    }
+
+    /// <inheritdoc cref="SolutionState.WithProjectAnalyzerReferences"/>
+    public SolutionCompilationState WithProjectAnalyzerReferences(
+        ProjectId projectId, IReadOnlyList<AnalyzerReference> analyzerReferences)
+    {
+        var stateChange = this.Solution.WithProjectAnalyzerReferences(projectId, analyzerReferences);
+        if (stateChange.NewSolutionState == this.Solution)
+            return this;
+
+        // The .Except() methods here aren't going to terribly cheap, but the assumption is adding or removing just the generators
+        // we changed, rather than creating an entire new generator driver from scratch and rerunning all generators, is cheaper
+        // in the end. This was written without data backing up that assumption, so if a profile indicates to the contrary,
+        // this could be changed.
+        //
+        // When we're comparing AnalyzerReferences, we'll compare with reference equality; AnalyzerReferences like AnalyzerFileReference
+        // may implement their own equality, but that can result in things getting out of sync: two references that are value equal can still
+        // have their own generator instances; it's important that as we're adding and removing references that are value equal that we
+        // still update with the correct generator instances that are coming from the new reference that is actually held in the project state from above.
+        // An alternative approach would be to call oldProject.WithAnalyzerReferences keeping all the references in there that are value equal the same,
+        // but this avoids any surprises where other components calling WithAnalyzerReferences might not expect that.
+
+        return ForkProject(
+            stateChange,
+            static stateChange =>
+            {
+                var addedReferences = stateChange.NewProjectState.AnalyzerReferences.Except<AnalyzerReference>(stateChange.OldProjectState.AnalyzerReferences, ReferenceEqualityComparer.Instance).ToImmutableArray();
+                var removedReferences = stateChange.OldProjectState.AnalyzerReferences.Except<AnalyzerReference>(stateChange.NewProjectState.AnalyzerReferences, ReferenceEqualityComparer.Instance).ToImmutableArray();
```

Comment:
At first this worried me since I was mentally imagining this lambda doesn't run until we end up pulling on the reslting project, i.e. we'd be holding onto the stateChange for awhile. But no -- it only avoids it if we skipped the "no change at all" check so it's fine, but hopefully nobody does try to hold onto more here.